### DISCUSSION TITLE
Any solution for db lock during transaction

### DISCUSSION THREAD

**Discussion by gaganbiswas** [pocketbase/pocketbase#5086] [Q&A]
```js
routerAdd("POST", "/update-stock", (c) => {
  const info = $apis.requestInfo(c);
  const { data } = info.data;

  const stock_collection = $app.dao().findCollectionByNameOrId("stock");

  $app.dao().runInTransaction((txDao) => {
    for (let item of data) {
      const stock_record = txDao.findFirstRecordByData(
        stock_collection.id,
        "product",
        item.product
      );

      const stock_form = new RecordUpsertForm($app, stock_record);
      stock_form.setDao(txDao);

      if (stock_record.get("id")) {
        stock_form.loadData({
          "quantity-": item.quantity,
        });
      }

      stock_form.submit();

      if (
        !txDao.canAccessRecord(stock_record, info, stock_collection.createRule)
      ) {
        throw new ForbiddenError("The operation is not allowed.");
      }
    }
  });

  return c.json(200, { status: 1 });
});
```

I have this code to update my stock collection. But since this is a transaction there is a db lock involved when updating the stocks. For my use case, multiple users are updating the stock collection. Is there any solution against this, maybe some other solution not involving the transaction and so that the update is processed accordingly and correctly even when multiple updates from different users are coming through.

Fields in collection:
`stock` --> id, product, quantity
`virtual_stock` --> id, product, quantity 

_Side Note: The application is a small stock management app which has a `virtual_stock` and a `stock` collection. Everytime an order is created the quantity of the products in the `virtual_stock` gets updated. And for delivery of an order the `stock` gets updated._

**Comment by ganigeorgiev** [comment]
Transactions (especially mixed read/write) require lock there is no workaround for this but note that even if there is a db lock PocketBase has auto fail-retry that will handle it behind the scenes. If you are not experiencing this, then something else is probably going on in your code that the above doesn't show.

Feel free to create a **minimal** reproducible repo and I will try to investigate it.

**Reply by gaganbiswas** [reply]
Thank you for the information. Let me investigate further into the functions before this transaction is called as you mentioned: 
>then something else is probably going on in your code that the above doesn't show.

If I am unable to spot anything, I will create a minimal reproducible repo and share it here.

**Reply by ganigeorgiev** [reply]
Also note that we have 2 separate db pools for read and write/transaction and for the latter there is  a limit to 1, so technically even if there are concurrent requests for the same transaction it should just queue up.
Deadlocks (not just SQLITE_BUSY or "database is locked" errors) could happen if you mix non-transactional writes inside a transaction (ex. if you didn't have `stock_form.setDao(txDao)`). But again the above code sample is not enough to guess what could cause the issue in your case.

**Reply by gaganbiswas** [reply]
Just two quick questions: 

1. Is there any way to call a route created in `pb` from another route. like say the above `/update-stock` from another route in `pb` say `/deliver-order`. 

2. Can I use the pocketbase hooks folder as my sole server instead of creating a separate server. For example, currently i am running some codes in from my frontend using `js_sdk` that modifies the pocketbase db and then calling this '/update-stock' to run the transaction and if the transaction fails i am running some more functions related to pocketbase in frontend using the `js_sdk`.

**Reply by gaganbiswas** [reply]
>Also note that we have 2 separate db pools for read and write/transaction and for the latter there is a limit to 1, so technically even if there are concurrent requests for the same transaction it should just queue up.
Deadlocks (not just SQLITE_BUSY or "database is locked" errors) could happen if you mix non-transactional writes inside a transaction (ex. if you didn't have stock_form.setDao(txDao)). But again the above code sample is not enough to guess what could cause the issue in your case.

Thanks. I will look into my code and test a bit more to identify the core problem. This cleared some parts where I was having doubts.

**Reply by ganigeorgiev** [reply]
1. No, unless you send a separate [http request with `$http.send`](https://pocketbase.io/docs/js-sending-http-requests/). But ideally you shouldn't need this and instead split the common parts between the 2 endpoints in a helper function (_note that the JSVM doesn't allow access to outside of the handler scope so you'll have to put it in module as shown in [Loading modules](https://pocketbase.io/docs/js-overview/#loading-modules)_).

2. I don't understand what do you mean by _" hooks folder as my sole server instead of creating a separate server"_. Also how the modification mentioned in _"i am running some codes in my server that modifies the pocketbase db"_ is performed?

**Reply by gaganbiswas** [reply]
>2. I don't understand what do you mean by _" hooks folder as my sole server instead of creating a separate server"_. Also how the modification mentioned in _"i am running some codes in my server that modifies the pocketbase db"_ is performed?

My bad for the confusion. Let me explain properly

When a order is getting created, two things should happen:
1. The order is to be created in the orders collection which has the following fields --> id: string, order_details: json, status: 0.
2. The `/update-stock` is called using `$http.send` which updates the quantity of the stock based on the order.

I am using the `js_sdk` in my frontend and am using the `create()` method to create the order in the orders collection and then using the `send()` method to hit the '/update-stock' with the order details to modify the quantity of the  products in the stock using a transaction _( the order details is an array of {product_id: string, quantity: number} )_.


My question is that can do both the things - creating the order and modifying the stock inside the '/update-stock'? so basically, using the custom routes created in the hooks folder of pocketbase as my server to modify the database.

**Reply by ganigeorgiev** [reply]
I'm sorry but I don't think I can help further. You've edited your replies several times and I no longer understand what is actually the issue here. Again, if you think that there is a PocketBase issue (I'm not ruling that out) feel free to create a **minimal** reproducible repo and I'll try to investigate it.

**Reply by gaganbiswas** [reply]
Not a problem. Thank you for the support. I will investigate further with the information you have provided.
